### PR TITLE
Add Mac Address Service

### PR DESCRIPTION
**Pull request by pk556**
# Testing Guide for MAC Address Service PR

## Overview
To ensure the new Mac Address Service routes are functioning correctly, follow these steps to test each route. Be aware of potential edge cases and ensure comprehensive coverage.

- `GET /macs/{mac}`
- `POST /macs`
- `PUT /macs/{mac}`

### Testing `GET /macs/{mac}`

**Purpose:** Retrieve a specific MAC address entity.

**Steps:**

1. **Basic Request:**
   - Send a GET request to `/macs/{mac}` with a valid MAC address.
   - Verify that the response status is `200 OK`.
   - Check that the response body contains the correct MAC address entity in JSON format.
   - **Example of Valid MAC Address**: `BC-D0-74-13-50-F0`

2. **MAC Address Format Testing:**
   - Test with various valid MAC address formats:
     - `001122334455`
     - `00:11:22:33:44:55`
     - `00-11-22-33-44-55`
   - Verify that all valid formats are accepted and return the correct entity.

3. **Invalid MAC Address:**
   - Test with invalid MAC addresses:
     - Incorrect format (e.g., `00:11:22:33:44:5`)
     - Invalid characters (e.g., `00:11:22:33:44:GG`)
   - Verify that the response status is `400 Bad Request` with an appropriate error message.

4. **Non-existent MAC Address:**
   - Test with a valid format but non-existent MAC address.
   - Verify that the response status is `404 Not Found` with an appropriate error message.

5. **Case Sensitivity:**
   - Test with mixed case MAC addresses (e.g., `00:aA:Bb:CC:dd:EE`).
   - Verify that the service handles case insensitively.

### Testing `POST /macs`

**Purpose:** Create a new MAC address entity.

**Steps:**

1. **Basic Request:**
   - Send a POST request to `/macs` with a valid JSON body:
     ```json
     {
       "mac": "001122334455",
       "macpool": 1,
       "properties": {
         "key1": "value1",
         "key2": "value2"
       }
     }
     ```
   - Verify that the response status is `200 OK`.
   - Check that the response body contains the object ID of the newly created MAC address.
   - **Example of valid PoolID**: `3610227`
   - Refer to [valid properties for MAC Address entities](https://docs.bluecatnetworks.com/r/Address-Manager-API-Guide/MAC-pool-objects/9.4.0) when passing in properties

2. **MAC Address Format Testing:**
   - Test creation with various valid MAC address formats (as in GET testing).
   - Verify that all valid formats are accepted.

3. **Invalid MAC Address:**
   - Test with invalid MAC addresses in the request body.
   - Verify that the response status is `400 Bad Request` with an appropriate error message.

4. **Missing Required Fields:**
   - Test with a request body missing the `mac` field.
   - Verify that the response status is `400 Bad Request` with an appropriate error message.

5. **Optional Fields:**
   - Test creation without the `macpool` field.
   - Test creation without the `properties` field.
   - Verify that the MAC address is created successfully in both cases.

6. **Duplicate MAC Address:**
   - Attempt to create a MAC address that already exists.
   - Verify that the service handles this appropriately (e.g., `409 Conflict` or updates the existing entry).

7. **Invalid Pool ID:**
   - Test with a non-existent `macpool` ID.
   - Verify that the service handles this error appropriately.

### Testing `PUT /macs/{mac}`

**Purpose:** Update an existing MAC address entity.

**Steps:**

1. **Basic Request:**
   - Send a PUT request to `/macs/{mac}` with a valid MAC address and JSON body:
     ```json
     {
       "macpool": 2,
       "properties": {
         "key3": "value3",
         "key4": "value4"
       }
     }
     ```
   - Verify that the response status is `204 No Content`.
   - **Example of valid PoolID**: `3610227`
   - Refer to [valid properties for MAC Address entities](https://docs.bluecatnetworks.com/r/Address-Manager-API-Guide/MAC-pool-objects/9.4.0) when passing in properties

2. **Updating Pool ID:**
   - Test updating only the `macpool` field.
   - Verify that the MAC address is associated with the new pool.

3. **Updating Properties:**
   - Test updating only the `properties` field.
   - Verify that new properties are added and existing ones are updated.

4. **Non-existent MAC Address:**
   - Attempt to update a non-existent MAC address.
   - Verify that the response status is `404 Not Found` with an appropriate error message.

5. **Invalid MAC Address Format:**
   - Test with an invalid MAC address in the URL.
   - Verify that the response status is `400 Bad Request` with an appropriate error message.

6. **Empty Update:**
   - Send an update request with an empty body.
   - Verify that the service handles this gracefully (e.g., no changes made, successful response).

7. **Partial Updates:**
   - Test updating only a subset of properties.
   - Verify that unmentioned properties remain unchanged.

### CODE REVIEW

File: internal/api/helpers.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		if err != nil {
+			// Log failure to write the response
+			logger.Error("Failed to write response", zap.Error(err))
+			http.Error(w, "Failed to write response", http.StatusInternalServerError)
+		}
+	}
+}
+
+// handleError handles standard apierror return codes
+func handleError(w http.ResponseWriter, err error) {
```
Comment (btassone) [inline internal/api/helpers.go:153]: This is never used btw

**Review by btassone — APPROVED** [review]
Looks good. I had to dif it from the dev-getEntities branch, and once I did I realized it was only like 5 files. Looks good to me!
